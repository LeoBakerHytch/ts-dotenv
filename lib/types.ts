export type Env = {
    [key: string]: string | undefined;
};

export type EnvSchemaValue =
    | BooleanConstructor
    | NumberConstructor
    | StringConstructor
    | RegExp
    | Array<string>;

export type EnvSchema = {
    [key: string]: EnvSchemaValue;
};

/**
 * Resolves to the type of the provided environment schema.
 *
 * @typeParam S - pass `typeof schema` here
 */
export type EnvType<S extends EnvSchema> = {
    [K in keyof S]: S[K] extends BooleanConstructor
        ? boolean
        : S[K] extends NumberConstructor
        ? number
        : S[K] extends StringConstructor
        ? string
        : S[K] extends RegExp
        ? string
        : never;
};
