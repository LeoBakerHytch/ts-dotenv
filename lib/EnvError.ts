import { EnvSchemaValue } from './types';

export enum EnvErrorType {
    MISSING = 'MISSING',
    WRONG_TYPE = 'WRONG_TYPE',
}

export type EnvErrorReport = {
    [key: string]: {
        type: EnvErrorType;
        schemaValue: EnvSchemaValue;
        receivedValue?: string;
    };
};

export class EnvError extends Error {
    readonly name: string;
    readonly report: EnvErrorReport;

    constructor(report: EnvErrorReport) {
        super(formatReport(report));
        this.name = 'EnvError';
        this.report = report;
    }
}

function formatReport(report: EnvErrorReport) {
    const errors = Object.entries(report).map(entry => {
        const [key, { type, schemaValue, receivedValue }] = entry;
        return formatError(key, type, schemaValue, receivedValue);
    });
    return `Invalid or missing environment variables\n    - ${errors.join('\n    - ')}\n`;
}

function formatError(
    key: string,
    type: EnvErrorType,
    schemaValue: EnvSchemaValue,
    value: string | undefined,
): string {
    switch (type) {
        case EnvErrorType.MISSING:
            return `Expected value for key '${key}'; none found`;

        case EnvErrorType.WRONG_TYPE:
            return schemaValue instanceof RegExp
                ? `Expected value for key '${key}' to match ${schemaValue}; got '${value}'`
                : `Expected value for key '${key}' of type ${schemaValue.name}; got '${value}'`;
    }
}
