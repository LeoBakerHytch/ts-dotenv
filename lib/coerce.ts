import { Env, EnvSchema, EnvType } from './types';

type CoercedEnv = {
    [key: string]: boolean | number | string;
};

export function coerce<S extends EnvSchema>(schema: S, env: Env): EnvType<S> {
    const coerced: CoercedEnv = {};

    for (const [key, schemaValue] of Object.entries(schema)) {
        const value = env[key] || '';

        if (schemaValue === Boolean) {
            coerced[key] = value === 'true';
        } else if (schemaValue === Number) {
            coerced[key] = parseInt(value, 10);
        } else {
            coerced[key] = value;
        }
    }

    return coerced as EnvType<S>;
}
