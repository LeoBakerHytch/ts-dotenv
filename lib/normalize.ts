import { EnvKeyConfig, EnvSchema, EnvSchemaValue, NormalizedSchema } from './types';

export function normalize(schema: EnvSchema): NormalizedSchema {
    const result: NormalizedSchema = {};
    for (const [key, schemaValue] of Object.entries(schema)) result[key] = getConfig(schemaValue);
    return result;
}

function getConfig(schemaValue: EnvSchemaValue): EnvKeyConfig {
    return 'type' in schemaValue
        ? {
              type: schemaValue.type,
              optional: 'optional' in schemaValue ? schemaValue.optional : true,
              default: 'default' in schemaValue ? schemaValue.default : undefined,
          }
        : {
              type: schemaValue,
              optional: false,
          };
}
