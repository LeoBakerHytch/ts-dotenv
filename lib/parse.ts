import { Env } from './types';

/**
 * @see http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html
 */
const lineRegExp = /^([a-zA-Z_]+[a-zA-Z0-9_]*)\s*=\s*(.*)$/;

/**
 * Accounts for CR / LF / CR+LF line breaks
 */
const lineBreaksRegExp = /\r\n|\r|\n/;

export function parse(dotEnv: string): Env {
    const lines = dotEnv.split(lineBreaksRegExp);
    const parsed: Env = {};

    for (const line of lines) {
        const match = line.trim().match(lineRegExp);
        if (!match) continue;

        const variableName = match[1];
        parsed[variableName] = match[2];
    }

    return parsed;
}
