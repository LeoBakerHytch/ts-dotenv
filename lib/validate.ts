import { EnvErrorReport, EnvErrorType, EnvError } from './EnvError';
import { Env, EnvSchema } from './types';

/**
 * Only allows exactly 'true' or 'false'
 */
const booleanRegExp = /^(true|false)$/;

/**
 * Only allows integers; crudely prevents values greater than MAX_SAFE_INTEGER.
 */
const numberRegExp = /^-?\d{1,15}$/;

export type ValidatedEnv = {
    [key: string]: string;
};

export function validate(schema: EnvSchema, env: Env): env is ValidatedEnv {
    const report = validateSchema(schema, env);
    if (report) throw new EnvError(report);
    return true;
}

function validateSchema(schema: EnvSchema, env: Env): EnvErrorReport | null {
    const report: EnvErrorReport = {};

    for (const [key, schemaValue] of Object.entries(schema)) {
        const value = env[key];

        if (!valueExists(value)) {
            report[key] = {
                type: EnvErrorType.MISSING,
                schemaValue,
            };
        } else if (!typeMatches(schemaValue, value)) {
            report[key] = {
                type: EnvErrorType.WRONG_TYPE,
                schemaValue,
                value,
            };
        }
    }

    return Object.values(report).some(value => value != null) ? report : null;
}

function valueExists(value: string | undefined): value is string {
    return !(value === undefined || value === '');
}

function typeMatches(schemaValue: any, value: string): boolean {
    switch (schemaValue) {
        case Boolean:
            return booleanRegExp.test(value);

        case Number:
            return numberRegExp.test(value);

        default:
            if (schemaValue instanceof RegExp) return schemaValue.test(value);
            return true;
    }
}
