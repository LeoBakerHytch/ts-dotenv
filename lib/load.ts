import { readFileSync } from 'fs';
import { parse } from './parse';
import { validate } from './validate';
import { EnvSchema, EnvType } from './types';
import { coerce } from './coerce';

interface Options {
    fileName?: string;
    encoding?: string;
    overrideProcessEnv?: boolean;
}

export function load<S extends EnvSchema>(schema: S, fileName?: string): EnvType<S>;
export function load<S extends EnvSchema>(schema: S, options?: Options): EnvType<S>;
export function load<S extends EnvSchema>(
    schema: S,
    fileNameOrOptions?: string | Options,
): EnvType<S> {
    const { fileName = '.env', encoding = 'utf-8', overrideProcessEnv = false } =
        typeof fileNameOrOptions === 'string'
            ? { fileName: fileNameOrOptions }
            : fileNameOrOptions || {};

    const raw = loadDotEnv(fileName, encoding);
    const parsed = parse(raw);

    const merged = overrideProcessEnv
        ? { ...process.env, ...parsed }
        : { ...parsed, ...process.env };

    if (validate(schema, merged)) {
        return coerce(schema, merged);
    }

    throw Error('Invalid or missing environment variables');
}

function loadDotEnv(fileName: string, encoding: string): string {
    try {
        return readFileSync(fileName, encoding);
    } catch {
        throw Error('Could not find environment file');
    }
}
