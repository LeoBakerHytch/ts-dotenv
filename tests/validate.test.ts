import { EnvError } from '../lib/EnvError';
import { validate } from '../lib/validate';

describe('validate', () => {
    describe('boolean', () => {
        const schema = {
            KEY: Boolean,
        };

        it('should allow true', () => {
            const env = { KEY: 'true' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should allow false', () => {
            const env = { KEY: 'false' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject a boolean with leading whitespace', () => {
            const env = { KEY: ' false' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });

        it('should reject a boolean with trailing whitespace', () => {
            const env = { KEY: 'false ' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });

        it('should reject any other string', () => {
            const env = { KEY: 'string' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });
    });

    describe('buffer', () => {
        const schema = {
            KEY: Buffer,
        };

        it('should allow a full word', () => {
            const env = { KEY: 'YWJj' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should allow a word padded with =', () => {
            const env = { KEY: 'YWI=' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should allow a word padded with ==', () => {
            const env = { KEY: 'YQ==' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject an improperly padded value', () => {
            const env = { KEY: 'YW=' };
            expect(() => validate(schema, env)).toThrow();
        });
    });

    describe('number', () => {
        const schema = {
            KEY: Number,
        };

        it('should allow positive integers', () => {
            const env = { KEY: '123' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should allow negative integers', () => {
            const env = { KEY: '-123' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should allow a large number', () => {
            const env = { KEY: '123456789012345' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject a number larger than the max safe integer', () => {
            const env = { KEY: '9007199254740992' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });
    });

    describe('regular expression', () => {
        const schema = {
            KEY: /abc/,
        };

        it('should allow a value that matches a provided regex', () => {
            const env = { KEY: 'abc' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject a value that does not match a provided regex', () => {
            const env = { KEY: 'xyz' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });
    });

    describe('string', () => {
        const schema = {
            KEY: String,
        };

        it('should allow strings', () => {
            const env = { KEY: 'abc 123' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject empty strings', () => {
            const env = { KEY: '' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });
    });

    describe('string union', () => {
        const schema = {
            KEY: ['value' as const],
        };

        it('should allow specified strings', () => {
            const env = { KEY: 'value' };
            expect(() => validate(schema, env)).not.toThrow();
        });

        it('should reject other strings', () => {
            const env = { KEY: 'another-value' };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });
    });

    describe('missing & extra keys', () => {
        const schema = {
            KEY: Boolean,
        };

        it('should reject an env missing a required key', () => {
            const env = {};
            expect(() => validate(schema, env)).toThrow(EnvError);
        });

        it('should reject an env with an undefined key', () => {
            const env = { KEY: undefined };
            expect(() => validate(schema, env)).toThrow(EnvError);
        });

        it('should allow an env with extra keys', () => {
            const env = { KEY: 'true', NODE_ENV: 'production' };
            expect(() => validate(schema, env)).not.toThrow();
        });
    });

    describe('optional keys', () => {
        it('should allow missing values for an optional key', () => {
            const schema = {
                KEY: {
                    type: Boolean,
                    optional: true,
                },
            } as const;
            expect(() => validate(schema, {})).not.toThrow();
        });

        it('should allow missing values for a key with a default', () => {
            const schema = {
                KEY: {
                    type: Boolean,
                    default: true,
                },
            };
            expect(() => validate(schema, {})).not.toThrow();
        });
    });

    describe('reporting', () => {
        it('should throw a custom error type', () => {
            const schema = {
                MISSING: Boolean,
            };

            try {
                validate(schema, {});
            } catch (error) {
                expect(error).toBeInstanceOf(EnvError);
                expect((error as EnvError).name).toEqual('EnvError');
            }
        });

        it('should report missing variables', () => {
            const schema = {
                MISSING: Boolean,
            };

            expect(() => validate(schema, {})).toThrowErrorMatchingSnapshot();
        });

        it('should report badly typed variables', () => {
            const schema = {
                BOOLEAN: Boolean,
                NUMBER: Number,
                REGEXP: /abc/,
                STRING_UNION: ['abc', 'def'],
            };

            const env = {
                BOOLEAN: '123',
                NUMBER: 'abc',
                REGEXP: 'true',
                STRING_UNION: 'xyz',
            };

            expect(() => validate(schema, env)).toThrowErrorMatchingSnapshot();
        });
    });
});
