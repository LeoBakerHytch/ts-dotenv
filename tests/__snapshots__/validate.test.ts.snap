// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate reporting should report badly typed variables 1`] = `
"EnvError: Invalid or missing environment variables
    - Expected value for key 'BOOLEAN' of type Boolean; got '123'
    - Expected value for key 'NUMBER' of type Number; got 'abc'
    - Expected value for key 'REGEXP' to match /abc/; got 'true'
"
`;

exports[`validate reporting should report missing variables 1`] = `
"EnvError: Invalid or missing environment variables
    - Expected value for key 'MISSING'; none found
"
`;
